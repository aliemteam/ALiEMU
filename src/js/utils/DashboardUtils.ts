import { Moment, unix } from 'moment';

type Users = ALiEMU.EducatorDashboard.UserObject;
type Courses = ALiEMU.EducatorDashboard.CourseData;
type User = ALiEMU.EducatorDashboard.UserMeta;
type CourseMeta = ALiEMU.EducatorDashboard.CourseMetaObject;
type Categories = ALiEMU.EducatorDashboard.CategoryObject;

/**
 * Either immediately triggers a file download or saves a downloadable file blob
 *   to an HTML anchor element, depending on browser support for the "download"
 *   attribute.
 *
 * @param {string} filename The name of the file (with .csv as the suffix).
 * @param {Blob}   blob     A prepared file blob for download.
 * @param {BrowserType} browser The browser type.
 * @param {string} ElementID The HTML ID of the target element.
 * @return {void}
 */
export function downloadPolyfill(
    filename: string,
    blob: Blob,
    browser: ALiEMU.BrowserType,
    ElementID: string
): void {
    switch (browser) {
        case 'chrome':
        case 'firefox':
        case 'edge':
            const a: HTMLAnchorElement = document.createElement('a');
            document.body.appendChild(a);
            a.href = URL.createObjectURL(blob);
            a.download = filename;
            a.click();
            document.body.removeChild(a);
            break;
        case 'ie':
            window.navigator.msSaveBlob(blob, filename);
            break;
        default:
            const target = document.getElementById(ElementID);
            if (!target) {
                return;
            }
            target.setAttribute('href', URL.createObjectURL(blob));
            target.setAttribute('download', filename);
            alert(
                'Right click the button and select "save target as" to download CSV file.'
            );
    }
}

/**
 * Retrieves the category for a given courseID
 * @param {string} courseID       The courseID.
 * @param {Categories} categories The entire CategoryObject.
 * @return {string}  The category.
 */
export function getCourseCategory(
    courseID: string,
    categories: Categories
): string | undefined {
    let category: string | undefined;
    Object.keys(categories).some(key => {
        if (categories[key].hasOwnProperty(courseID)) {
            category = key;
            return true;
        }
        return false;
    });
    return category;
}

/**
 * Returns either a formatted date or 'X' based on whether or not the date string
 *   is defined.
 *
 * @param date  The date string (or undefined).
 * @param hours The associated hours for the course.
 * @return The parsed date and hours string or X","X. (format allows array join)
 */
export function parseCompletionData(
    date: number | undefined,
    hours?: string
): string {
    if (!hours) {
        if (date === undefined) {
            return `X`;
        }
        return unix(date).calendar();
    }

    if (date === undefined) {
        return `X","0`;
    }
    return `${unix(date).calendar()}","${hours}`;
}

/**
 * Calculates a given user's III hours based on the recommended hours for the course.
 *
 * @param  {User} user A given user's metadata object.
 * @param  {CourseMeta} courseMeta The complete CourseMetaObject.
 * @return {number} The user's calculated III hours.
 */
export function calculateIIIHours(
    user: User,
    courseMeta: CourseMeta,
    dateRange: { start: Moment | null; end: Moment | null }
): number {
    if (!user.courseCompleted) return 0;
    return Object.keys(user.courseCompleted).reduce((prev, curr) => {
        const { start, end } = dateRange;
        const completionDate = (<{[id: string]: number}>user.courseCompleted)![curr];
        const d = unix(completionDate);
        const addedHours: number =
            parseFloat((<any>courseMeta)[curr].recommendedHours) + prev;
        switch (true) {
            case !start && !end:
                return addedHours;
            case !start && end !== null:
                if (d.isBefore(end!) || d.isSame(end!)) return addedHours;
                return prev;
            case start !== null && !end:
                if (d.isAfter(start!) || d.isSame(start!)) return addedHours;
                return prev;
            default:
                const isBetween = d.isBetween(start!, end!);
                const isSame = d.isSame(start!) || d.isSame(end!);
                if (isBetween || isSame) return addedHours;
                return prev;
        }
    }, 0);
}

export class CSV {
    private users: Users;
    private courses: ALiEMU.EducatorDashboard.CourseObject;
    private courseData: ALiEMU.EducatorDashboard.CourseData;
    private courseMeta: ALiEMU.EducatorDashboard.CourseMetaObject;
    private lessons: ALiEMU.EducatorDashboard.Lessons;

    constructor(users: Users, courses: Courses) {
        this.users = users;
        this.courses = courses.courses;
        this.courseData = courses;
        this.courseMeta = courses.courseMeta;
        this.lessons = courses.lessons;
    }

    allUsers(dateRange: ALiEMU.EducatorDashboard.DateRange): ALiEMU.CSV {
        const filename = 'ALiEMU_Program_Export.csv';
        let data =
            [
                'Last Name',
                'First Name',
                'Class of',
                'Total III Hours Awarded',
                'Courses In Progress',
                'Courses Completed',
            ]
                .map(i => `"${i}"`)
                .join(',') + '\n';

        Object.keys(this.users).forEach((uid: string) => {
            const id = parseInt(uid, 10);
            const inProgress: number = this.users[id].courseProgress
                ? Object.keys(this.users[id].courseProgress!).length
                : 0;

            const completed: number = this.users[id].courseCompleted
                ? Object.keys(this.users[id].courseCompleted!).length
                : 0;
            data +=
                [
                    this.users[id].lastName,
                    this.users[id].firstName,
                    this.users[id].auGraduationYear || '',
                    calculateIIIHours(
                        this.users[id],
                        this.courseMeta,
                        dateRange
                    ),
                    inProgress - completed,
                    completed,
                ]
                    .map(i => `"${i}"`)
                    .join(',') + '\n';
        });
        return { filename, data };
    }

    user(userID: string): ALiEMU.CSV | boolean {
        const id = parseInt(userID, 10);
        const courseProgress = this.users[id].courseProgress;
        const { courses, courseMeta, categories } = this.courseData;

        if (!courseProgress) return false;

        const filename = `${this.users[id].displayName.replace(
            /\s/,
            '_'
        )}.csv`;
        let data: string =
            [
                'Registered Courses',
                'Steps Completed',
                'Date Completed',
                'Associated III Credit Hours',
                'Category',
            ]
                .map(i => `"${i}"`)
                .join(',') + '\n';

        for (const key of Object.keys(courseProgress)) {
            const k = parseInt(key, 10);
            data +=
                [
                    courses[k].postTitle,
                    `${courseProgress[k].completed} out of ${courseProgress[
                        k
                    ].total}`,
                    parseCompletionData(
                        (<any>this.users[id]).courseCompleted![k],
                        (<any>courseMeta[k]).recommendedHours
                    ),
                    getCourseCategory(key, categories),
                ]
                    .map(i => `"${i}"`)
                    .join(',') + '\n';
        }
        return { filename, data };
    }

    course(courseID: string): ALiEMU.CSV {
        const id = parseInt(courseID, 10);
        const filename = `${this.courses[id].postTitle.replace(
            /\s/,
            '_'
        )}.csv`;
        const lessonIDs: string[] = [];

        let data: string =
            [
                'Last Name',
                'First Name',
                'Course Completed',
                ...this.courses[id].lessons
                    .filter(
                        (lessonID: string) =>
                            this.lessons[parseInt(lessonID, 10)] !== undefined
                    )
                    .map((lessonID: string) => {
                        lessonIDs.push(lessonID);
                        return `Lesson: ${this.lessons[parseInt(lessonID, 10)].postTitle}`;
                    }),
            ]
                .map(i => `"${i}"`)
                .join(',') + '\n';

        Object.keys(this.users).forEach((userID: string) => {
            const uid = parseInt(userID, 10);
            const completionData = this.users[uid].courseCompleted
                ? parseCompletionData(
                      (<any>this.users[uid]).courseCompleted![courseID]
                  )
                : 'X';
            data +=
                [
                    this.users[uid].lastName,
                    this.users[uid].firstName,
                    completionData,
                    ...lessonIDs.map((lessonID: string) => {
                        try {
                            const completed = this.users[uid].courseProgress![
                                id
                            ].lessons[parseInt(lessonID, 10)];
                            if (completed === 1) {
                                return 'Completed';
                            }
                            return 'X';
                        } catch (e) {
                            return 'X';
                        }
                    }),
                ]
                    .map(i => `"${i}"`)
                    .join(',') + '\n';
        });
        return { filename, data };
    }
}
