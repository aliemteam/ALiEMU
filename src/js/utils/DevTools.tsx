// tslint:disable:no-typeof-undefined
import DevTool from 'mobx-react-devtools';
import * as React from 'react';

export default function devtool(props: any): JSX.Element | null {
    if (process.env.NODE_ENV !== 'production') {
        return <DevTool {...props} />;
    }
    return null;
}

export function configureDevtool(options: {
    logEnabled?: boolean;
    updatesEnabled?: boolean;
    graphEnabled?: boolean;
    logFilter?: (p: any) => boolean;
}): void {
    if (process.env.NODE_ENV !== 'production') {
        const cdt = require('mobx-react-devtools').configureDevtool;
        cdt(options);
    }
}
