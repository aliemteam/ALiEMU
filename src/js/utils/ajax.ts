interface AjaxSuccess<T> {
    success: true;
    data: T;
}

interface AjaxError {
    success: false;
    data: {
        code: string;
        message: string;
    };
}

type AjaxResponse<T> = AjaxSuccess<T> | AjaxError;

const { AU_AJAX } = window;

export async function wpAjax<T>(
    action: string,
    data: { [k: string]: Scalar },
): Promise<AjaxResponse<T>> {
    const { nonce, url } = AU_AJAX;
    let response: AjaxResponse<T>;
    try {
        response = await jQuery.post(url, {
            _ajax_nonce: nonce,
            action,
            ...data,
        });
    } catch (_e) {
        response = {
            success: false,
            data: {
                code: 'http_error',
                message:
                    'An unexpected error occurred while handling your request. Please try again later.',
            },
        };
    }
    return response;
}
