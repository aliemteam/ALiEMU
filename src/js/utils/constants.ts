/**
 * Used to set intent-based styles on various components across the project.
 */
export const enum Intent {
    PRIMARY = 'primary',
    SECONDARY = 'secondary',
    SUCCESS = 'success',
    WARNING = 'warning',
    DANGER = 'danger',
}

/**
 * Course status enum used in `User.course_progress`.
 */
export const enum CourseStatus {
    COMPLETED = 'COMPLETED',
    STARTED = 'STARTED',
}

export const UserTitles: ReadonlyArray<string> = [
    'Physician',
    'Pharmacist',
    'Physician Assistant',
    'Nurse Practitioner',
    'Nurse',
    'Physical Therapist',
    'Occupational Therapist',
    'Prehospital Provider',
    'Respiratory Therapist',
    'Other',
];

export const UserPracticeLevels: ReadonlyArray<string> = [
    'Student',
    'Resident',
    'Fellow',
    'Non-Trainee Practitioner',
];
