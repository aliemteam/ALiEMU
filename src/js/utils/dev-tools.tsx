import DevTool from 'mobx-react-devtools';
import * as React from 'react';

export default (): JSX.Element | null => {
    if (process.env.NODE_ENV !== 'production') {
        return <DevTool position={{ bottom: 0, right: 0 }} />;
    }
    return null;
};

interface ConfigureOptions {
    logEnabled?: boolean;
    updatesEnabled?: boolean;
    graphEnabled?: boolean;
    logFilter?: (p: any) => boolean;
}

export const configureDevtool = (options: ConfigureOptions): void => {
    if (process.env.NODE_ENV !== 'production') {
        const cdt = require('mobx-react-devtools').configureDevtool;
        cdt(options);
    }
};
