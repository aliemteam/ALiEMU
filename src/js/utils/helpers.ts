/**
 * Small polyfill / helper for extracting URL search params as a Map.
 */
export function mapUrlParams(): Map<string, string> {
    const searchParams = window.location.search;
    if ('URLSearchParams' in window) {
        return new Map([...new URLSearchParams(searchParams).entries()]);
    } else {
        return new Map(
            searchParams
                .substring(1)
                .split('&')
                .map(
                    (item): [string, string] => {
                        const [key, val] = item.split('=');
                        return [
                            key,
                            decodeURIComponent(val).replace(/\+/g, ' '),
                        ];
                    },
                ),
        );
    }
}
