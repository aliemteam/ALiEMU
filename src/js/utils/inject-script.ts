async function waitForProperty(propertyKey: string): Promise<void> {
    const propertyChain = propertyKey.split('.');
    // tslint:disable-next-line:promise-must-complete
    await new Promise((resolve, reject) => {
        let iterations = 0;
        const interval = setInterval(() => {
            let property: any = window;
            let propertyChainSuccess = true;
            for (const key of propertyChain) {
                if (!property.hasOwnProperty(key)) {
                    propertyChainSuccess = false;
                    break;
                }
                property = property[key];
            }
            if (propertyChainSuccess) {
                clearInterval(interval);
                resolve();
            }
            if (iterations > 10) {
                reject();
            }
            iterations++;
        }, 500);
    });
}

function createScriptManager(baseElement: HTMLElement) {
    const registered: Set<string> = new Set();
    return async (src: string, propertyKey?: string): Promise<void> => {
        if (!registered.has(src)) {
            const script = document.createElement('script');
            script.src = src;
            script.async = true;
            script.defer = true;
            await new Promise(resolve => {
                script.addEventListener('load', resolve);
                baseElement.appendChild(script);
            });
            if (propertyKey) {
                await waitForProperty(propertyKey);
            }
            registered.add(src);
        }
    };
}

export default createScriptManager(document.body);
