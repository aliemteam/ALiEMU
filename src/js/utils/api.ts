import { ICoach, ICourse, ILearner } from 'utils/types';

interface Params {
    [k: string]: string | number | boolean | string[] | number[];
}

type IComment = WordPress.Comment;
type IUser = WordPress.User<'edit'>;

export class Comments {
    // {{{
    private static endpoint = '/wp-json/wp/v2/comments';

    static async fetchOne(id: number, params?: Params): Promise<IComment> {
        return _get(`${Comments.endpoint}/${id}`, params).then(data =>
            data.json(),
        );
    }
    // }}}
}

export class Courses {
    // {{{
    private static endpoint = '/wp-json/aliemu/v1/courses';

    static async fetchOne(id: number, params?: Params): Promise<ICourse> {
        return _get(`${Courses.endpoint}/${id}`, params).then(res =>
            res.json(),
        );
    }

    static async fetchMany(
        params: Params = {},
        startPage = 1,
        endPage = Infinity,
    ): Promise<ICourse[]> {
        return _getMany<ICourse>(Courses.endpoint, params, startPage, endPage);
    }
    // }}}
}

export class Groups {
    // {{{
    private static endpoint = '/wp-json/wp/v2/users/me/groups';

    static async fetchCoaches(params: Params = {}): Promise<ICoach[]> {
        return _getMany<ICoach>(`${Groups.endpoint}/coaches`, params);
    }

    static async addCoach(email: string): Promise<ICoach> {
        return _post<ICoach>(`${Groups.endpoint}/coaches`, { email });
    }

    static async removeCoach(id: number): Promise<Response> {
        return _delete(`${Groups.endpoint}/coaches/${id}`);
    }

    static async fetchLearners(params: Params = {}): Promise<ILearner[]> {
        return _getMany<ILearner>(`${Groups.endpoint}/learners`, params);
    }

    static async removeLearner(id: number): Promise<Response> {
        return _delete(`${Groups.endpoint}/learners/${id}`);
    }

    static async addLearnerTag(id: number, tag: string): Promise<string[]> {
        return _post<string[]>(`${Groups.endpoint}/learners/${id}/tags`, {
            tag,
        });
    }

    static async removeLearnerTag(id: number, tag: string): Promise<Response> {
        return _delete(`${Groups.endpoint}/learners/${id}/tags`, { tag });
    }
    // }}}
}

export class Users {
    // {{{
    private static endpoint = '/wp-json/wp/v2/users';

    static async update(fields: Partial<IUser> | FormData, id?: number): Promise<IUser> {
        return _post<IUser>(`${Users.endpoint}/${id ? id : 'me'}`, fields);
    }
    // }}}
}

function _get(endpoint: string, params?: Params): Promise<Response> {
    // {{{
    if (params) {
        endpoint = endpoint + parseParams(params);
    }
    return fetch(endpoint, {
        headers: {
            'X-WP-Nonce': window.AU_API.nonce,
        },
        mode: 'same-origin',
    });
    // }}}
}

function _post<T>(endpoint: string, params: object | FormData): Promise<T> {
    // {{{
    return fetch(endpoint, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
            'X-WP-Nonce': window.AU_API.nonce,
        },
        body: JSON.stringify(params),
        mode: 'same-origin',
    }).then(data => data.json());
    // }}}
}

function _delete(endpoint: string, params: object = {}): Promise<Response> {
    // {{{
    return fetch(endpoint, {
        method: 'DELETE',
        headers: {
            'Content-Type': 'application/json',
            'X-WP-Nonce': window.AU_API.nonce,
        },
        body: JSON.stringify(params),
        mode: 'same-origin',
    });
    // }}}
}

function parseParams(params: Params): string {
    // {{{
    return (
        '?' +
        [...Object.entries(params)]
            .reduce(
                (arr, [key, val]) => [
                    ...arr,
                    `${key}=${
                        Array.isArray(val)
                            ? encodeURIComponent(val.join(','))
                            : val
                    }`,
                ],
                <string[]>[],
            )
            .join('&')
    );
    // }}}
}

async function _getMany<T>(
    endpoint: string,
    params: Params = {},
    page: number = 1,
    endPage = Infinity,
    collection: T[] = [],
): Promise<T[]> {
    // {{{
    params = {
        ...params,
        page,
    };
    const response = await _get(endpoint, params);

    if (!response.headers.has('X-WP-TotalPages')) {
        throw new Error(
            'X-WP-TotalPages header must be set to iterate entire collection.',
        );
    }
    const totalPages = parseInt(response.headers.get('X-WP-TotalPages')!, 10);

    collection = [...collection, ...(await response.json())];

    return page >= totalPages || page === endPage
        ? collection
        : _getMany(endpoint, params, page + 1, endPage, collection);
    // }}}
}

// vim: set fdm=marker:
