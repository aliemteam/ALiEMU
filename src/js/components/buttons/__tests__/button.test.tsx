import React from 'react';
import renderer from 'react-test-renderer';

import { Intent } from '../../../utils/constants';
import Button from '../button';

it('should render correctly', () => {
    const tree = renderer.create(
        <>
            <Button>Primary</Button>
            <Button intent={Intent.SUCCESS}>Success</Button>
            <Button intent={Intent.WARNING}>Warning</Button>
            <Button intent={Intent.DANGER}>Danger</Button>
            <Button loading>Loading</Button>
        </>,
    );
    expect(tree).toMatchSnapshot();
});
