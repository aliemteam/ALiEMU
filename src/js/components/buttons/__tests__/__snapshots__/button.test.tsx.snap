// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
Array [
  <button
    className="button"
    onClick={[Function]}
  >
    <span>
      Primary
    </span>
  </button>,
  <button
    className="button intentSuccess"
    onClick={[Function]}
  >
    <span>
      Success
    </span>
  </button>,
  <button
    className="button intentWarning"
    onClick={[Function]}
  >
    <span>
      Warning
    </span>
  </button>,
  <button
    className="button intentDanger"
    onClick={[Function]}
  >
    <span>
      Danger
    </span>
  </button>,
  <button
    className="button loading"
    disabled={true}
    onClick={[Function]}
  >
    <span>
      Loading
    </span>
    <div
      className="spinner"
    >
      <svg
        className="spinner"
        height={25}
        style={
          Object {
            "strokeDasharray": "70.68583470577035px",
            "strokeDashoffset": "56.54866776461628px",
            "strokeWidth": "2.5px",
          }
        }
        viewBox="0 0 25 25"
        width={25}
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx={12.5}
          cy={12.5}
          r={11.25}
        />
      </svg>
    </div>
  </button>,
]
`;
